### PR TITLE
Serialization improvements step 4 (undo.h)

### PR DESCRIPTION
**Pull request by sipa**
The next step of changes from #10785.

This one adds:
* A meta-formatter for vectors, which serializes the vector elements using another formatter
* Switch the undo.h code to the new framework, using the above (where undo entries are serialized as a vector, each of which uses a modified serializer for the UTXOs).

### CODE REVIEW

**Comment by sipa** [comment]
@ryanofsky I added another small improvement here that wasn't present in #10785, namely reusing the `VectorUsing` logic in the vector/prevector serializers directly (avoiding code duplication).

File: src/serialize.h [hunk 1]
```diff
@@ -25,6 +25,9 @@
 
 static const unsigned int MAX_SIZE = 0x02000000;
 
+/** Maximum amount of memory to allocate at once when deserializing vectors. */
```
Comment (Empact) [inline src/serialize.h]: nit: specify units?

Comment (sipa) [reply src/serialize.h]: Done.

**Comment by kallewoof** [comment]
Code view ACK 24235e50e9d532cbe5836dcebbce0c6a04be8652

File: src/serialize.h [hunk 2]
```diff
@@ -673,6 +723,16 @@ inline void Unserialize(Stream& is, T&& a)
     a.Unserialize(is);
 }
 
+/** Identity formatter. Serializes objects as themselves. */
```
Comment (laanwj) [inline src/serialize.h]: Maybe add a comment when this is to be used, on first glance it seems like a no-op.

File: src/serialize.h [hunk 3]
```diff
@@ -593,6 +596,53 @@ class LimitedString
...
+    void Unser(Stream& s, V& v)
+    {
+        v.clear();
+        size_t size = ReadCompactSize(s);
+        size_t allocated = 0;
+        while (allocated < size) {
+            // For DoS prevention, do not blindly allocate as much as the stream claims to contain.
+            // Instead, allocate in 5MiB batches, so that an attacker actually needs to provide
+            // X MiB of data to make us allocate X+5 Mib.
+            allocated = std::min(size, allocated + MAX_VECTOR_ALLOCATE / sizeof(typename V::value_type));
```
Comment (laanwj) [inline src/serialize.h]: A static assert that ` sizeof(typename V::value_type) <= MAX_VECTOR_ALLOCATE` could catch an edge case here.

File: src/serialize.h [hunk 4]
```diff
@@ -596,6 +596,53 @@ class LimitedString
...
+ *     SERIALIZE_METHODS(X, obj) { READWRITE(Using<VectorUsing<VarInt>>(obj.v)); }
+ *   };
+ * will define a struct that contains a vector of uint64_t, which is serialized
+ * as a vector of VarInt-encoded integers.
+ *
+ * V is not required to be an std::vector type. It works for any class that
+ * exposes a value_type, size, reserve, push_back, and const iterators.
+ */
+template<class Formatter>
+struct VectorUsing
```
Comment (ryanofsky) [inline src/serialize.h]: I don't think the `VectorUsing` and `UsingSelf` names for formatters are the best. Would suggest sticking to the `FooFormatter` convention and renaming `VectorUsing` to `VectorFormatter` and `UsingSelf` to `DefaultFormatter`.

Aside from giving more consistent class names, it should also make usages more readable: `Using<VectorUsing<UsingSelf>>(v)` -> `Using<VectorFormatter<DefaultFormatter>>(v)`

**Review by ryanofsky — APPROVED** [review]
Code review ACK 24235e50e9d532cbe5836dcebbce0c6a04be8652. Left minor suggestions, feel free to ignore

File: src/serialize.h (hunk 2 above)
Comment (sipa) [reply src/serialize.h]: Done.

File: src/serialize.h (hunk 3 above)
Comment (sipa) [reply src/serialize.h]: Done.

File: src/serialize.h [hunk 5]
```diff
@@ -596,6 +596,53 @@ class LimitedString
...
+            // For DoS prevention, do not blindly allocate as much as the stream claims to contain.
+            // Instead, allocate in 5MiB batches, so that an attacker actually needs to provide
+            // X MiB of data to make us allocate X+5 Mib.
+            allocated = std::min(size, allocated + MAX_VECTOR_ALLOCATE / sizeof(typename V::value_type));
+            v.reserve(allocated);
+            while (v.size() < allocated) {
+                typename V::value_type val;
+                auto wrapped = Using<Formatter>(val);
+                s >> wrapped;
+                v.push_back(std::move(val));
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/serialize.h]: Done.

File: src/undo.h [hunk 6]
```diff
@@ -13,88 +13,50 @@
...
-    }
-
-    template <typename Stream>
-    void Unserialize(Stream& s) {
-        // TODO: avoid reimplementing vector deserializer
-        uint64_t count = 0;
-        ::Unserialize(s, COMPACTSIZE(count));
-        if (count > MAX_INPUTS_PER_BLOCK) {
-            throw std::ios_base::failure("Too many input undo records");
-        }
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/undo.h]: Yeah, I'm aware. I don't think this is significant. The earlier code needed a sanity limit, and had custom code, so it used a custom limit. The new code is safe by default (even though the limit in practice is much higher).

File: src/serialize.h (hunk 4 above)
Comment (sipa) [reply src/serialize.h]: That's indeed better. Done.

File: src/serialize.h [hunk 7]
```diff
@@ -593,6 +596,53 @@ class LimitedString
...
+    template<typename Stream, typename V>
+    void Unser(Stream& s, V& v)
+    {
+        v.clear();
+        size_t size = ReadCompactSize(s);
+        size_t allocated = 0;
+        while (allocated < size) {
+            // For DoS prevention, do not blindly allocate as much as the stream claims to contain.
+            // Instead, allocate in 5MiB batches, so that an attacker actually needs to provide
+            // X MiB of data to make us allocate X+5 Mib.
```
Comment (jonatack) [inline src/serialize.h]: Lines 632 and 633, so the documentation doesn't go out of date if the value of `MAX_VECTOR_ALLOCATE` changes, perhaps replace "5 MiB" and "5 Mib" with "MAX_VECTOR_ALLOCATE" (and s/X Mib/X/ as the constant is defined in bytes).

File: src/serialize.h [hunk 8]
```diff
@@ -593,6 +596,53 @@ class LimitedString
...
+            static_assert(sizeof(typename V::value_type) <= MAX_VECTOR_ALLOCATE, "Vector element size too large");
+            allocated = std::min(size, allocated + MAX_VECTOR_ALLOCATE / sizeof(typename V::value_type));
```
Comment (jonatack) [inline src/serialize.h]: Now that the static assert suggested in https://github.com/bitcoin/bitcoin/pull/18021#discussion_r375166767 has been added, possibly worthwhile (or not) to calculate `sizeof(typename V::value_type)` once and cache it.

File: src/serialize.h [hunk 9]
```diff
@@ -593,6 +596,53 @@ class LimitedString
...
+    void Ser(Stream& s, const V& v)
+    {
+        WriteCompactSize(s, v.size());
+        for (const typename V::value_type& elem : v) {
+            s << Using<Formatter>(elem);
+        }
+    }
+
+    template<typename Stream, typename V>
+    void Unser(Stream& s, V& v)
```
Comment (jonatack) [inline src/serialize.h]: nit: With respect to the naming of `Ser` and `Unser`, git grepping for only these might be facilitated with names that don't correspond to the first chars of Serialize/Unserialize. That said, I don't have a suggestion top of mind.

**Review by jonatack — COMMENTED** [review]
Qualified ACK 3c94b0039d2c

Code review, built/ran tests/ran bitcoind. I am not yet highly competent wrt C++ templates but this looks like a worthwhile refactoring. Would review again. Some thoughts below, feel free to ignore.

File: src/serialize.h (hunk 9 above)
Comment (jonatack) [reply src/serialize.h]: Never mind, `git grep "Ser(" and `git grep "Unser("` does the job.

File: src/serialize.h (hunk 8 above)
Comment (jonatack) [reply src/serialize.h]: Noting that the result of sizeof is always nonzero, even if applied to an empty class type; so no risk of division by zero.

Comment (kallewoof) [reply src/serialize.h]: Unless I'm confused, sizeof(..) is determined at compile-time, so no reason to cache it.

Comment (jonatack) [reply src/serialize.h]: Oops yes, `static_assert` requires a bool constexpr, otherwise a compile-time error is issued.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 3c94b0039d2ca2a8c41fd6127ff5019a2afc304e. Changes since last review: renaming formatter classes, adding suggested static_assert, and removing temporary in VectorFormatter

**Comment by laanwj** [comment]
code review ACK 3c94b0039d2ca2a8c41fd6127ff5019a2afc304e

**Pull request merged**